src/networking.c

```diff
@@ -3543,14 +3582,14 @@ void flushSlavesOutputBuffers(void) {
  * The function always succeed, even if there is already a pause in progress.
  * In such a case, the duration is set to the maximum and new end time and the
  * type is set to the more restrictive type of pause. */
-void pauseClients(mstime_t end, pause_type type) {
-    if (type > server.client_pause_type) {
-        server.client_pause_type = type;
-    }
-
-    if (end > server.client_pause_end_time) {
-        server.client_pause_end_time = end;
+void pauseClients(pause_purpose purpose, mstime_t end, pause_type type) {
+    /* Manage pause type and end time per pause purpose. */
+    if (server.client_pause_per_purpose[purpose] == NULL) {
+        server.client_pause_per_purpose[purpose] = zmalloc(sizeof(pause_event));
     }
+    server.client_pause_per_purpose[purpose]->type = type;
+    server.client_pause_per_purpose[purpose]->end = end;
```

Comment:
at least for CLIENT PAUSE command, we wanted to take the max end time and most restrictive mode if called twice.
it's a bit odd that CLIENT UNPAUSE cancels both, but i think acceptable.

IIRC FAILOVER is rejected if one is already in progress.
and SHUTDOWN can just silently join the existing shutdown (or we can merge the timeouts, max or min, i don't care)

Reply:
:facepalm: you're right. I wanted (but forgot...) to set them to max(old, new) for all purposes, although it only matters for CLIENT PAUSE.